<a href="https://colab.research.google.com/github/FedeM-Git/TT-2C2025-Data-Analitycs-Notebooks/blob/main/Pre-Entrega_Maune/Pre-Entrega_Maune%2C%20Juan_Federico.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>


<img  src="https://raw.githubusercontent.com/FedeM-Git/TT-2C2025-Data-Analitycs-Notebooks/refs/heads/main/Pre-Entrega_Maune/Pre-Entrega%20IMG.PNG" width=100%>

#**MAUNE, Juan Federico**
#**Pre-Entrega de Proyecto**

**Bases de datos a utilizar:**
- [clientes.csv](https://github.com/FedeM-Git/TT-2C2025-Data-Analitycs-Notebooks/blob/main/Pre-Entrega_Maune/datasets/clientes.csv)
- [marketing.csv](https://github.com/FedeM-Git/TT-2C2025-Data-Analitycs-Notebooks/blob/main/Pre-Entrega_Maune/datasets/marketing.csv)
- [ventas.csv](https://github.com/FedeM-Git/TT-2C2025-Data-Analitycs-Notebooks/blob/main/Pre-Entrega_Maune/datasets/ventas.csv)


#**Etapa 1: Recopilación y preparación de Datos**

##***Actividad 1: Carga de los sets de datos***

In [1]:
# se importa pandas
import pandas as pd

In [2]:
# se importa data set clientes
df_clientes = pd.read_csv("https://raw.githubusercontent.com/FedeM-Git/TT-2C2025-Data-Analitycs-Notebooks/refs/heads/main/Pre-Entrega_Maune/datasets/clientes.csv")

In [3]:
# se importa data set marketing
df_marketing = pd.read_csv("https://raw.githubusercontent.com/FedeM-Git/TT-2C2025-Data-Analitycs-Notebooks/refs/heads/main/Pre-Entrega_Maune/datasets/marketing.csv")

In [4]:
# se importa data set ventas
df_ventas = pd.read_csv("https://raw.githubusercontent.com/FedeM-Git/TT-2C2025-Data-Analitycs-Notebooks/refs/heads/main/Pre-Entrega_Maune/datasets/ventas.csv")

##***Actividad 2 y 3: Script Básico***

In [ ]:
# se paso parte de los registros del set de datos de "Ventas" a una lista de diccionarios
lista_ventas = [
{'id_venta': 2824, 'producto': 'Tablet', 'precio': 74.09, 'cantidad': 5, 'fecha_venta': '45294', 'categoria': 'Electrónica'},
{'id_venta': 446, 'producto': 'Cortinas', 'precio': 62.10, 'cantidad': 4, 'fecha_venta': '45295', 'categoria': 'Decoración'},
{'id_venta': 630, 'producto': 'Adorno de pared', 'precio': 71.99, 'cantidad': 9, 'fecha_venta': '45295', 'categoria': 'Decoración'},
{'id_venta': 794, 'producto': 'Jarrón decorativo', 'precio': 96.97, 'cantidad': 7, 'fecha_venta': '45295', 'categoria': 'Decoración'},
{'id_venta': 882, 'producto': 'Cuadro decorativo', 'precio': 101.92, 'cantidad': 9, 'fecha_venta': '45295', 'categoria': 'Decoración'},
{'id_venta': 969, 'producto': 'Candelabro', 'precio': 90.70, 'cantidad': 12, 'fecha_venta': '45295', 'categoria': 'Decoración'},
{'id_venta': 1445, 'producto': 'Lavadora', 'precio': 28.41, 'cantidad': 10, 'fecha_venta': '45295', 'categoria': 'Electrodomésticos'},
{'id_venta': 1971, 'producto': 'Freidora eléctrica', 'precio': 122.25, 'cantidad': 11, 'fecha_venta': '45295', 'categoria': 'Electrodomésticos'},
{'id_venta': 2209, 'producto': 'Proyector', 'precio': 32.15, 'cantidad': 1, 'fecha_venta': '45295', 'categoria': 'Electrónica'},
{'id_venta': 256, 'producto': 'Cortinas', 'precio': 86.82, 'cantidad': 8, 'fecha_venta': '45296', 'categoria': 'Decoración'},
{'id_venta': 606, 'producto': 'Cortinas', 'precio': 106.66, 'cantidad': 11, 'fecha_venta': '45296', 'categoria': 'Decoración'},
{'id_venta': 704, 'producto': 'Jarrón decorativo', 'precio': 57.22, 'cantidad': 9, 'fecha_venta': '45296', 'categoria': 'Decoración'},
{'id_venta': 1362, 'producto': 'Heladera', 'precio': 109.49, 'cantidad': 11, 'fecha_venta': '45296', 'categoria': 'Electrodomésticos'},
{'id_venta': 2155, 'producto': 'Auriculares', 'precio': 84.80, 'cantidad': 10, 'fecha_venta': '45296', 'categoria': 'Electrónica'},
{'id_venta': 2947, 'producto': 'Cámara digital', 'precio': 107.78, 'cantidad': 3, 'fecha_venta': '45296', 'categoria': 'Electrónica'},
{'id_venta': 1495, 'producto': 'Lavadora', 'precio': 38.99, 'cantidad': 6, 'fecha_venta': '45297', 'categoria': 'Electrodomésticos'},
{'id_venta': 1572, 'producto': 'Cafetera', 'precio': 61.40, 'cantidad': 9, 'fecha_venta': '45297', 'categoria': 'Electrodomésticos'},
{'id_venta': 2180, 'producto': 'Consola de videojuegos', 'precio': 102.57, 'cantidad': 9, 'fecha_venta': '45297', 'categoria': 'Electrónica'},
{'id_venta': 2487, 'producto': 'Cámara digital', 'precio': 50.76, 'cantidad': 3, 'fecha_venta': '45297', 'categoria': 'Electrónica'},
{'id_venta': 2603, 'producto': 'Smartphone', 'precio': 87.05, 'cantidad': 5, 'fecha_venta': '45297', 'categoria': 'Electrónica'},
{'id_venta': 2885, 'producto': 'Auriculares', 'precio': 100.40, 'cantidad': 6, 'fecha_venta': '45297', 'categoria': 'Electrónica'},
{'id_venta': 17, 'producto': 'Elementos de cerámica', 'precio': 108.48, 'cantidad': 11, 'fecha_venta': '45298', 'categoria': 'Decoración'},
{'id_venta': 183, 'producto': 'Espejo decorativo', 'precio': 82.76, 'cantidad': 8, 'fecha_venta': '45298', 'categoria': 'Decoración'},
{'id_venta': 560, 'producto': 'Adorno de pared', 'precio': 57.62, 'cantidad': 6, 'fecha_venta': '45298', 'categoria': 'Decoración'},
{'id_venta': 632, 'producto': 'Cuadro decorativo', 'precio': 30.49, 'cantidad': 12, 'fecha_venta': '45298', 'categoria': 'Decoración'}
]

In [ ]:
# ventas acumuladas de la categoría "Decoración"
ventas_decoracion = sum( [venta ['precio']* venta['cantidad']
                          for venta in lista_ventas
                          if venta['categoria']=='Decoración'])
print(f'El monto obtenido por las ventas en la categoría Decoración es de $ {ventas_decoracion}')

El monto obtenido por las ventas en la categoría Decoración es de $ 8530.539999999999


In [ ]:
# ventas acumuladas de la categoría "Electrodométicos"
ventas_electrodomesticos = sum( [venta ['precio']* venta['cantidad']
                                 for venta in lista_ventas
                                 if venta['categoria']=='Electrodomésticos'])
print(f'El monto obtenido por las ventas en la categoría Electrodomésticos es de $ {ventas_electrodomesticos}')

El monto obtenido por las ventas en la categoría Electrodomésticos es de $ 3619.7799999999997


In [ ]:
# ventas acumuladas de la categoría "Electrónica"
ventas_electronica = sum( [venta ['precio']* venta['cantidad']
                           for venta in lista_ventas
                           if venta['categoria']=='Electrónica'])
print(f'El monto obtenido por las ventas en la categoría Electrónica es de $ {ventas_electronica}')

El monto obtenido por las ventas en la categoría Electrónica es de $ 3687.0


In [ ]:
# utilicé la AI para obtener un código que unifique los codigos anteriores en una sola sentencia
ventas_por_categoria = {
    categoria: sum(venta['precio'] * venta['cantidad']
                   for venta in lista_ventas if venta['categoria'] == categoria)
    for categoria in {venta['categoria'] for venta in lista_ventas}
}

print(ventas_por_categoria)

{'Electrónica': 3687.0, 'Decoración': 8530.539999999999, 'Electrodomésticos': 3619.7799999999997}


## ***Actividad 4: Calidad de los Datos***
**Análisis exploratorio con Pandas**

### ***DataFrame Clientes***

#### **Análisis Exploratorio**

In [5]:
# Realizo un paneo del dataset mediante info() para conocer columnas, filas, tipos de datos y si hay datos nulos
df_clientes.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 567 entries, 0 to 566
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   id_cliente  567 non-null    int64  
 1   nombre      567 non-null    object 
 2   edad        567 non-null    int64  
 3   ciudad      567 non-null    object 
 4   ingresos    567 non-null    float64
dtypes: float64(1), int64(2), object(2)
memory usage: 22.3+ KB


In [7]:
# obtengo una muestra medinte sample() para ver los datos en la tabla y ver si hay coherencia en la misma.
df_clientes.sample(10)

id_cliente               nombre  edad                 ciudad  ingresos
417         418  Melita Watson-Brown    27  San Miguel de Tucumán  22533.03
122         123   Mattie Humpherston    35  San Miguel de Tucumán  23801.33
12           13     Hadria Hinkensen    33  San Miguel de Tucumán  56171.38
221         222          Bary Tanner    37                Córdoba  37626.61
512         513       Brian Longland    24                Posadas  16211.05
223         224    Dudley Stallworth    33           Buenos Aires  20196.84
308         309       Frieda Sappell    42  San Miguel de Tucumán  37477.47
533         534   Graham Cuthbertson    36            Resistencia  48565.08
57           58     Phillipp Litster    45                Posadas  30978.96
380         381   Hartwell Balharrie    43            Resistencia  26112.90

In [8]:
# utilizo describe() para obtener estadísticas de las columnas numéricas y corroborar los rangos de los datos
df_clientes.describe()

id_cliente        edad      ingresos
count  567.000000  567.000000    567.000000
mean   284.000000   37.940035  34668.739012
std    163.823075   10.202885  12974.531446
min      1.000000   20.000000    170.290000
25%    142.500000   30.000000  26015.240000
50%    284.000000   37.000000  35066.830000
75%    425.500000   43.000000  42457.100000
max    567.000000   81.000000  88053.010000

In [9]:
# utilizo value_count() sobre la columna ciudad para ver si hay ciudades cargadas de diferente forma
df_clientes['ciudad'].value_counts()

ciudad
Mar del Plata            63
Rosario                  55
Posadas                  52
Resistencia              50
Córdoba                  49
Corrientes               47
Santa Fe                 46
Bahía Blanca             44
Salta                    43
Merlo                    43
San Miguel de Tucumán    39
Buenos Aires             36
Name: count, dtype: int64

In [12]:
# ordenos los nombres en orden ascendente para ver si hay nombres que inician con números o simbolos
df_clientes['nombre'].sort_values(ascending=True).head(20)

519          Aarika Brehat
403             Ab Wilsher
349        Abbey Vanichkin
184            Addy Lattos
103        Adolphus Morton
333         Adrianna Fenby
96                Ag Ifill
67         Agustin Bofield
536          Aidan Leavens
327        Ailina Martelet
70          Aindrea Winter
361    Albrecht Itchingham
245         Aldous Stivers
458     Alethea Jacobovitz
237      Alexandra Stapley
544             Ali Roxbee
529            Alicea Clow
63          Alison Ianizzi
148       Allissa Napleton
0          Aloysia Screase
Name: nombre, dtype: object

In [17]:
# use AI para obtener una sentencia que verifique la presencia de números o simbolos en la columna nombres
df_clientes[df_clientes["nombre"].str.contains(r"[^a-zA-ZáéíóúÁÉÍÓÚñÑ\s]", na=False)]

#Debio a que en la salida aprarecienron nombres ingles que contienen guien medio(-), apóstrofe (') y puntos (.), las entencia se cambió a :

#df_clientes[df_clientes["nombre"].str.contains(r"[^a-zA-ZáéíóúÁÉÍÓÚñÑ\s\.\-']", na=False)]

id_cliente                nombre  edad                 ciudad  ingresos
29           30          Tedi O'Shiel    32           Bahía Blanca  44137.90
124         125        Bennie O'Regan    40            Resistencia  60645.26
153         154    My de'-Ancy Willis    43                Rosario  33105.01
156         157   Westbrook O'Caherny    25               Santa Fe  29982.55
159         160    Holly-anne Whymark    43          Mar del Plata  30588.42
187         188          D'arcy Creak    42           Buenos Aires  36868.84
246         247      Shaw Dell'Abbate    33                Posadas  41806.83
268         269      Carilyn O'Mailey    52           Bahía Blanca   3603.65
290         291  Marie-jeanne Johnson    28               Santa Fe  42821.99
294         295   Bette-ann Bertholin    40           Bahía Blanca  35319.17
417         418   Melita Watson-Brown    27  San Miguel de Tucumán  22533.03
445         446       Gillan St. Hill    29          Mar del Plata  52757.36
466         467     Felipe O'Hallagan    31             Corrientes  32928.99
472         473         Dollie O'Roan    25            Resistencia  51350.49
545         546     Roseline L'oiseau    39             Corrientes  44987.46

In [ ]:
# corroboro la auscencia de valores nulos
df_clientes.isnull().sum()

id_cliente    0
nombre        0
edad          0
ciudad        0
ingresos      0
dtype: int64

In [ ]:
# exploro la presencia de registros duplicados
df_clientes.duplicated().sum()

np.int64(0)

####**Observaciones**


- cantidad de registros: 567
- columnas: 5
- Precisión:
    - los datos estan en los rangos esperados
- Validez:
    - los datos poseen los tipos de datos correctos
    - los nombres no poseen números ni símbolos
- Completitud:
    - no se observan registros con valores nulos
- Consistencia:
    - las ciudades están escritas de igual forma
    - no hay registros duplicados
- **Cambios a realizar:**
    - no se obserban

### ***DataFrame Marketing***

#### **Análisis Exploratorio**

In [ ]:
# Realizo un paneo del dataset mediante info() para conocer columnas, filas, tipos de datos y si hay datos nulos
df_marketing.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 90 entries, 0 to 89
Data columns (total 6 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   id_campanha   90 non-null     int64  
 1   producto      90 non-null     object 
 2   canal         90 non-null     object 
 3   costo         90 non-null     float64
 4   fecha_inicio  90 non-null     object 
 5   fecha_fin     90 non-null     object 
dtypes: float64(1), int64(1), object(4)
memory usage: 4.3+ KB


In [18]:
# obtengo una muestra medinte sample() para ver los datos en la tabla y ver si hay coherencia en la misma.
df_marketing.sample(10)

id_campanha                producto  canal  costo fecha_inicio   fecha_fin
42            7               Proyector   RRSS   5.57   30/07/2024  16/08/2024
81           59               Televisor  Email   4.19   20/11/2024   19/1/2025
12           50  Consola de videojuegos  Email   4.97   03/05/2024  28/05/2024
31           37               Proyector  Email   3.22   06/07/2024  26/07/2024
61           24     Parlantes Bluetooth   RRSS   6.00   22/09/2024  26/10/2024
15           46                Lavadora  Email   3.63   12/05/2024  29/05/2024
74           27                Batidora   RRSS   4.52   01/11/2024  08/12/2024
82           25              SmartWatch   RRSS   4.84   23/11/2024  29/12/2024
22            4                Heladera   RRSS   4.53   31/05/2024  15/07/2024
13           34                Heladera  Email   4.20   06/05/2024  05/07/2024

In [19]:
# utilizo describe() para obtener estadísticas de las columnas numéricas y corroborar los rangos de los datos
df_marketing.describe()

id_campanha      costo
count    90.000000  90.000000
mean     45.500000   4.928667
std      26.124701   0.947750
min       1.000000   2.950000
25%      23.250000   4.372500
50%      45.500000   4.900000
75%      67.750000   5.562500
max      90.000000   7.390000

In [23]:
# utilizo value_count() sobre la columna "producto" para ver si hay un mismo producto está cargado con diferentes nombres
df_marketing['producto'].sort_values().value_counts()

producto
Adorno de pared           3
Alfombra                  3
Aspiradora                3
Auriculares               3
Batidora                  3
Cafetera                  3
Candelabro                3
Consola de videojuegos    3
Cortinas                  3
Cuadro decorativo         3
Cámara digital            3
Elementos de cerámica     3
Espejo decorativo         3
Freidora eléctrica        3
Heladera                  3
Horno eléctrico           3
Jarrón decorativo         3
Laptop                    3
Lavadora                  3
Lámpara de mesa           3
Microondas                3
Parlantes Bluetooth       3
Plancha de vapor          3
Proyector                 3
Rincón de plantas         3
Secadora                  3
SmartWatch                3
Smartphone                3
Tablet                    3
Televisor                 3
Name: count, dtype: int64

In [24]:
# utilizo value_count() sobre la columna "canal" para ver si un canal está cargado con diferentes nombres
df_marketing['canal'].sort_values().value_counts()

canal
Email    30
RRSS     30
TV       30
Name: count, dtype: int64

In [25]:
# corroboro la auscencia de valores nulos
df_marketing.isnull().sum()

id_campanha     0
producto        0
canal           0
costo           0
fecha_inicio    0
fecha_fin       0
dtype: int64

In [26]:
# exploro la presencia de registros duplicados
df_marketing.duplicated().sum()

np.int64(0)

####**Observaciones**

- cantidad de registros: 90
- columnas: 6
- Precisión:
    - los datos estan en los rangos esperados
    - los registro de fechas no pueden ser analizado
- Validez:
    - las columnas 'fecha_inicio' y 'fecha_fin' tiene con Dtype Object y están en formato *dd/mm/aaaa*
- Completitud:
    - no se observan registros con valores nulos
- Consistencia:
    - las productos y canales están escritas de igual forma
    - no hay registros duplicados
- **Cambios a realizar:**
    1. cambiar el Dtype de 'fecha_inicio' y 'fecha_fin' a *datetime*

### ***DataFrame Ventas***

#### **Análisis Exploratorio**

In [27]:
# Realizo un paneo del dataset mediante info() para conocer columnas, filas, tipos de datos y si hay datos nulos
df_ventas.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3035 entries, 0 to 3034
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   id_venta     3035 non-null   int64  
 1   producto     3035 non-null   object 
 2   precio       3033 non-null   object 
 3   cantidad     3033 non-null   float64
 4   fecha_venta  3035 non-null   object 
 5   categoria    3035 non-null   object 
dtypes: float64(1), int64(1), object(4)
memory usage: 142.4+ KB


In [29]:
# obtengo una muestra medinte sample() para ver los datos en la tabla y ver si hay coherencia en la misma.
df_ventas.sample(10)

id_venta            producto   precio  cantidad fecha_venta  \
2614       889     Lámpara de mesa   $95.87       7.0  09/11/2024   
460       1323          Microondas   $52.26       8.0  23/02/2024   
630       1772            Heladera   $56.94       2.0  14/03/2024   
310       1855            Lavadora  $123.28       8.0  05/02/2024   
1498       625            Alfombra  $124.91       3.0  18/06/2024   
2355       924   Jarrón decorativo   $64.39      11.0  05/10/2024   
2367       474   Jarrón decorativo   $38.87      10.0  07/10/2024   
1450      2235         Auriculares   $74.81      12.0  11/06/2024   
438       1031  Freidora eléctrica  $117.97       6.0  21/02/2024   
443       1820    Plancha de vapor   $45.50      12.0  21/02/2024   

              categoria  
2614         Decoración  
460   Electrodomésticos  
630   Electrodomésticos  
310   Electrodomésticos  
1498         Decoración  
2355         Decoración  
2367         Decoración  
1450        Electrónica  
438   Electrodomésticos  
443   Electrodomésticos

In [44]:
# utilizo describe() para obtener estadísticas de las columnas numéricas y corroborar los rangos de los datos
df_ventas.describe()

id_venta     cantidad
count  3035.000000  3033.000000
mean   1499.851400     6.496538
std     866.465379     3.457250
min       1.000000     1.000000
25%     748.500000     3.000000
50%    1502.000000     7.000000
75%    2249.500000     9.000000
max    3000.000000    12.000000

In [35]:
# utilizo value_count() sobre la columna "productos" para ver si un producto está cargado con diferentes nombres
df_ventas['producto'].value_counts()

producto
Lámpara de mesa           181
Auriculares               147
Microondas                135
Cafetera                  118
Laptop                    104
Secadora                  103
Jarrón decorativo         102
Rincón de plantas         102
Smartphone                102
SmartWatch                102
Aspiradora                102
Alfombra                  102
Cortinas                  102
Adorno de pared           101
Freidora eléctrica        101
Espejo decorativo         101
Plancha de vapor          100
Cámara digital            100
Consola de videojuegos    100
Batidora                  100
Cuadro decorativo         100
Elementos de cerámica     100
Televisor                 100
Proyector                 100
Horno eléctrico           100
Lavadora                  100
Parlantes Bluetooth        99
Tablet                     58
Heladera                   49
Candelabro                 24
Name: count, dtype: int64

In [36]:
# utilizo value_count() sobre la columna "categoria" para ver si una categoria esta cargada con diferentes nombres
df_ventas['categoria'].value_counts()

categoria
Decoración           1015
Electrónica          1012
Electrodomésticos    1008
Name: count, dtype: int64

In [37]:
# corroboro la auscencia de valores nulos
df_ventas.isnull().sum()

id_venta       0
producto       0
precio         2
cantidad       2
fecha_venta    0
categoria      0
dtype: int64

In [38]:
# visualizo los volares nulos
df_ventas[df_ventas.isnull().any(axis=1)]

id_venta               producto precio  cantidad fecha_venta    categoria
136       627  Elementos de cerámica    NaN       NaN  17/01/2024   Decoración
139      2171    Parlantes Bluetooth    NaN       NaN  17/01/2024  Electrónica

In [42]:
# exploro la presencia de registros duplicados
df_ventas.duplicated().sum()

np.int64(35)

In [43]:
# visualizo los valores duplicados
df_ventas[df_ventas.duplicated(keep=False)].sort_values(by="id_venta")

id_venta           producto  precio  cantidad fecha_venta    categoria
820        56           Cortinas  $66.24       5.0  05/04/2024   Decoración
826        56           Cortinas  $66.24       5.0  05/04/2024   Decoración
832       145           Alfombra  $41.72       8.0  06/04/2024   Decoración
844       145           Alfombra  $41.72       8.0  06/04/2024   Decoración
878       204  Jarrón decorativo  $94.75       1.0  08/04/2024   Decoración
..        ...                ...     ...       ...         ...          ...
830      2545        Auriculares  $32.81      11.0  05/04/2024  Electrónica
883      2705        Auriculares  $87.29       2.0  08/04/2024  Electrónica
889      2705        Auriculares  $87.29       2.0  08/04/2024  Electrónica
831      2778         SmartWatch  $70.66       7.0  05/04/2024  Electrónica
825      2778         SmartWatch  $70.66       7.0  05/04/2024  Electrónica

[70 rows x 6 columns]

####**Observaciones**


precio (object) --> float64

fecha_venta (object) --> datetime

cantidad (float64) --> int64

los precios inclyen el simbolo ($), hay que quitarlos

no obsevo valores nulos con info, hay 2 registros con valres nulos  (0.07% de los registros)

hay 35 valores duplicados

- cantidad de registros: 3035
- columnas: 6
- Precisión:
    - los datos de *'cantidad'* están en rangos esperables. No hay valores negativos
    - no se puede indicar la precisión de *'precio'*, *'fecha_venta'* por tener el Dtype object
- Validez:
    - las columnas *'precio'* y *'fecha_venta'* poseen el Dtype Object. *'precio'* posee el prefijo '$'. *'fecha_venta'* está en formato *dd/mm/aaaa*
- Completitud:
    - se observa 2 registros con valores nulos (2/3035 => 0.07% de los registros)
- Consistencia:
    - los valores de categoría y productos están escritos de igual forma
    - hay 35 registros duplicados
- **Cambios a realizar:**
    1. eliminar los 2 registros con datos nulos
    2. eliminar los registros duplicados
    3. eliminar el prefijo '$' de los datos de *'precio'*
    4. cambiar Dtype de *'precio'* a *float64*
    5. cambiar Dtype de *'fecha_venta'* a *datetime*

#**Etapa 2: Preprocesamiento y Limpieza de Datos**

##***Actividad 1: Limpieza de Datos***

### ***DataFrame Marketing***

- **Cambios a realizar:**
    1. cambiar el Dtype de 'fecha_inicio' y 'fecha_fin' a *datetime*

In [45]:
# corroboro estado inicial del dataframe
df_marketing.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 90 entries, 0 to 89
Data columns (total 6 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   id_campanha   90 non-null     int64  
 1   producto      90 non-null     object 
 2   canal         90 non-null     object 
 3   costo         90 non-null     float64
 4   fecha_inicio  90 non-null     object 
 5   fecha_fin     90 non-null     object 
dtypes: float64(1), int64(1), object(4)
memory usage: 4.3+ KB


In [48]:
# se convierte el campo fecha_inicio a dtype datetime
df_marketing['fecha_inicio'] = pd.to_datetime(df_marketing['fecha_inicio'], format="%d/%m/%Y")

In [47]:
# se convierte el campo fecha_inicio a dtype datetime
df_marketing['fecha_fin'] = pd.to_datetime(df_marketing['fecha_fin'], format="%d/%m/%Y")

In [49]:
# corroboro los cambios de Dtype realizados
df_marketing.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 90 entries, 0 to 89
Data columns (total 6 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   id_campanha   90 non-null     int64         
 1   producto      90 non-null     object        
 2   canal         90 non-null     object        
 3   costo         90 non-null     float64       
 4   fecha_inicio  90 non-null     datetime64[ns]
 5   fecha_fin     90 non-null     datetime64[ns]
dtypes: datetime64[ns](2), float64(1), int64(1), object(2)
memory usage: 4.3+ KB


### ***DataFrame Ventas***

- **Cambios a realizar:**
    1. eliminar los 2 registros con datos nulos
    2. eliminar los registros duplicados
    3. eliminar el prefijo '$' de los datos de *'precio'*
    4. cambiar Dtype de *'precio'* a *float64*
    5. cambiar Dtype de *'fecha_venta'* a *datetime*

####*1. Eliminación de registros con datos nulos*

In [50]:
# corroboro presencia de registros con valores null
df_ventas[df_ventas.isnull().any(axis=1)]

id_venta               producto precio  cantidad fecha_venta    categoria
136       627  Elementos de cerámica    NaN       NaN  17/01/2024   Decoración
139      2171    Parlantes Bluetooth    NaN       NaN  17/01/2024  Electrónica

In [51]:
# Elimino los registros con valores nulos. Aplico los cambios en el mismo DataFrame
df_ventas.dropna(inplace=True)

In [52]:
# corroboro la eliminación de los registros con valores null
df_ventas[df_ventas.isnull().any(axis=1)]

Empty DataFrame
Columns: [id_venta, producto, precio, cantidad, fecha_venta, categoria]
Index: []

In [54]:
# corroboro que tengo 2 registro menos en el DataFrame Ventas
df_ventas.info()

<class 'pandas.core.frame.DataFrame'>
Index: 3033 entries, 0 to 3034
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   id_venta     3033 non-null   int64  
 1   producto     3033 non-null   object 
 2   precio       3033 non-null   object 
 3   cantidad     3033 non-null   float64
 4   fecha_venta  3033 non-null   object 
 5   categoria    3033 non-null   object 
dtypes: float64(1), int64(1), object(4)
memory usage: 165.9+ KB


####*2. Eliminación de registros duplicados*

In [53]:
# corroboro la presencia de los registros duplicados ordanados por id_venta
df_ventas[df_ventas.duplicated(keep=False)].sort_values(by="id_venta")

id_venta           producto  precio  cantidad fecha_venta    categoria
820        56           Cortinas  $66.24       5.0  05/04/2024   Decoración
826        56           Cortinas  $66.24       5.0  05/04/2024   Decoración
832       145           Alfombra  $41.72       8.0  06/04/2024   Decoración
844       145           Alfombra  $41.72       8.0  06/04/2024   Decoración
878       204  Jarrón decorativo  $94.75       1.0  08/04/2024   Decoración
..        ...                ...     ...       ...         ...          ...
830      2545        Auriculares  $32.81      11.0  05/04/2024  Electrónica
883      2705        Auriculares  $87.29       2.0  08/04/2024  Electrónica
889      2705        Auriculares  $87.29       2.0  08/04/2024  Electrónica
831      2778         SmartWatch  $70.66       7.0  05/04/2024  Electrónica
825      2778         SmartWatch  $70.66       7.0  05/04/2024  Electrónica

[70 rows x 6 columns]

In [55]:
# elimino los registros duplicados, aplico los cambios en el mismo DataFrame y reinico el indice del DataFrame
df_ventas.drop_duplicates(inplace=True, ignore_index=True)

In [57]:
# corroboro la eliminacionde los registros duplicados
df_ventas[df_ventas.duplicated(keep=False)].sort_values(by="id_venta")

Empty DataFrame
Columns: [id_venta, producto, precio, cantidad, fecha_venta, categoria]
Index: []

In [59]:
# vuelvo a ver el número de registros en el DataFrame
df_ventas.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2998 entries, 0 to 2997
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   id_venta     2998 non-null   int64  
 1   producto     2998 non-null   object 
 2   precio       2998 non-null   object 
 3   cantidad     2998 non-null   float64
 4   fecha_venta  2998 non-null   object 
 5   categoria    2998 non-null   object 
dtypes: float64(1), int64(1), object(4)
memory usage: 140.7+ KB


####*3. Eliminación el prefijo '$' de los datos de 'precio'*

In [61]:
# corroboro la presencia del prefijo '$' en la columna 'precio'
df_ventas.sample(10)

id_venta                producto  precio  cantidad fecha_venta  \
2856       759         Lámpara de mesa  $82.40      11.0  14/12/2024   
1433      2740  Consola de videojuegos  $88.36       5.0  13/06/2024   
1577       360         Adorno de pared  $67.02       9.0  04/07/2024   
115       2098              SmartWatch  $75.65       5.0  15/01/2024   
1291      1226                Secadora  $89.37      10.0  27/05/2024   
2237      2248              SmartWatch  $88.46       8.0  25/09/2024   
330        367   Elementos de cerámica  $47.00       2.0  08/02/2024   
2695      2479               Proyector  $51.72       3.0  23/11/2024   
1981       412       Cuadro decorativo  $67.79       9.0  24/08/2024   
2781      2140  Consola de videojuegos  $91.09       7.0  02/12/2024   

              categoria  
2856         Decoración  
1433        Electrónica  
1577         Decoración  
115         Electrónica  
1291  Electrodomésticos  
2237        Electrónica  
330          Decoración  
2695        Electrónica  
1981         Decoración  
2781        Electrónica

In [62]:
# Elimino los posibles espacios en blanco y Reeplazo el simbolo '$' por nada
df_ventas["precio"] = df_ventas["precio"].str.strip().str.replace("$", "")

In [63]:
# corroboro la ausencia del prefijo '$' en la columna 'precio'
df_ventas.sample(10)

id_venta               producto  precio  cantidad fecha_venta  \
2122      1490       Plancha de vapor   57.93       6.0  10/09/2024   
338       1371     Freidora eléctrica   95.71       4.0  09/02/2024   
853       1225               Lavadora   87.65       2.0  09/04/2024   
180        547  Elementos de cerámica   83.81       6.0  22/01/2024   
2391      1178               Cafetera   61.69       8.0  14/10/2024   
1533      2919              Proyector   46.95       6.0  28/06/2024   
2205       441        Lámpara de mesa  120.13      11.0  21/09/2024   
1912      1894             Aspiradora   55.65       5.0  15/08/2024   
448        928      Rincón de plantas  111.80       5.0  22/02/2024   
379        406               Cortinas   64.96       1.0  14/02/2024   

              categoria  
2122  Electrodomésticos  
338   Electrodomésticos  
853   Electrodomésticos  
180          Decoración  
2391  Electrodomésticos  
1533        Electrónica  
2205         Decoración  
1912  Electrodomésticos  
448          Decoración  
379          Decoración

####*4. Modificación del Dtype de *'precio'* a *float64**

In [64]:
# corroboro estado inicial del dataframe
df_ventas.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2998 entries, 0 to 2997
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   id_venta     2998 non-null   int64  
 1   producto     2998 non-null   object 
 2   precio       2998 non-null   object 
 3   cantidad     2998 non-null   float64
 4   fecha_venta  2998 non-null   object 
 5   categoria    2998 non-null   object 
dtypes: float64(1), int64(1), object(4)
memory usage: 140.7+ KB


In [65]:
# convierto los valore de la columna 'precio' al Dtype float64
df_ventas['precio'] = df_ventas['precio'].astype('float64')

In [66]:
# corroboro el cambio de Dtype de 'precio'
df_ventas.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2998 entries, 0 to 2997
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   id_venta     2998 non-null   int64  
 1   producto     2998 non-null   object 
 2   precio       2998 non-null   float64
 3   cantidad     2998 non-null   float64
 4   fecha_venta  2998 non-null   object 
 5   categoria    2998 non-null   object 
dtypes: float64(2), int64(1), object(3)
memory usage: 140.7+ KB


####*5. Modificación del Dtype de *'fecha_venta'* a *datetime**

In [67]:
# corroboro estado inicial del dataframe
df_ventas.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2998 entries, 0 to 2997
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   id_venta     2998 non-null   int64  
 1   producto     2998 non-null   object 
 2   precio       2998 non-null   float64
 3   cantidad     2998 non-null   float64
 4   fecha_venta  2998 non-null   object 
 5   categoria    2998 non-null   object 
dtypes: float64(2), int64(1), object(3)
memory usage: 140.7+ KB


In [68]:
# se convierte el campo fecha_venta a dtype datetime
df_ventas['fecha_venta'] = pd.to_datetime(df_ventas['fecha_venta'], format="%d/%m/%Y")

In [69]:
# corroboro el cambio de Dtype de 'fecha_venta'
df_ventas.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2998 entries, 0 to 2997
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype         
---  ------       --------------  -----         
 0   id_venta     2998 non-null   int64         
 1   producto     2998 non-null   object        
 2   precio       2998 non-null   float64       
 3   cantidad     2998 non-null   float64       
 4   fecha_venta  2998 non-null   datetime64[ns]
 5   categoria    2998 non-null   object        
dtypes: datetime64[ns](1), float64(2), int64(1), object(2)
memory usage: 140.7+ KB


##***Actividad 2: Transformacion de Datos***

Aplicar filtros y transformaciones para crear una tabla de ventas que muestre solo los productos con alto rendimiento

In [70]:
# se crea una nueva columna 'monto_venta' = 'precio'*'cantidad'
df_ventas['monto_venta'] = df_ventas['precio']*df_ventas['cantidad']

In [71]:
# visualizo la creación de la nuevo columna
df_ventas.sample(15)

id_venta               producto  precio  cantidad fecha_venta  \
1201      1006               Secadora  116.90       5.0  2024-05-17   
2596      1497        Horno eléctrico   83.30      10.0  2024-11-11   
2917      2451    Parlantes Bluetooth   42.08       2.0  2024-12-20   
1746      2709              Proyector  108.23       6.0  2024-07-25   
1756       290        Adorno de pared  100.50      12.0  2024-07-27   
2468      1733             Microondas   33.38      11.0  2024-10-26   
34        2603             Smartphone   87.05       5.0  2024-01-06   
2220      1782               Heladera   99.66      12.0  2024-09-23   
1630      1884             Aspiradora   94.69       4.0  2024-07-11   
199       2543             Smartphone   91.39       7.0  2024-01-24   
1455       534      Jarrón decorativo   32.38       5.0  2024-06-17   
721        708      Rincón de plantas   63.10      11.0  2024-03-24   
1598       827  Elementos de cerámica   97.35       5.0  2024-07-07   
739        649        Lámpara de mesa   83.61       1.0  2024-03-26   
1835      2114                 Tablet   64.99       3.0  2024-08-04   

              categoria  monto_venta  
1201  Electrodomésticos       584.50  
2596  Electrodomésticos       833.00  
2917        Electrónica        84.16  
1746        Electrónica       649.38  
1756         Decoración      1206.00  
2468  Electrodomésticos       367.18  
34          Electrónica       435.25  
2220  Electrodomésticos      1195.92  
1630  Electrodomésticos       378.76  
199         Electrónica       639.73  
1455         Decoración       161.90  
721          Decoración       694.10  
1598         Decoración       486.75  
739          Decoración        83.61  
1835        Electrónica       194.97

In [75]:
# creo un nuevo DataFrame agrupando los productos y sumando los montos de las ventas
df_tabla_Ventas = df_ventas.groupby(["producto"], as_index=False)["monto_venta"].sum()

In [76]:
# corroboro la creación del nuevo DataFrame
df_tabla_Ventas.head()

producto  monto_venta
0  Adorno de pared     48093.49
1         Alfombra     44773.06
2       Aspiradora     50085.86
3      Auriculares     74175.58
4         Batidora     50979.20

In [77]:
# corroboro métricas estadísticas de la nueva tabla
df_tabla_Ventas.describe()

monto_venta
count     30.000000
mean   48903.117333
std    13265.561715
min    11128.800000
25%    45601.920000
50%    48140.245000
75%    51092.960000
max    82276.380000

In [78]:
# corroboro fórmula para obtener el Q3
df_tabla_Ventas['monto_venta'].quantile(0.75)

np.float64(51092.96)

In [80]:
# creo un nuevo DataFrame con Filtro los datos para quedarme con los productos de cuartil superior (Q3)
df_productosTop=df_tabla_Ventas[df_tabla_Ventas['monto_venta']>df_tabla_Ventas['monto_venta'].quantile(0.75)]

In [81]:
# corroboro el nuevo DataFrame con los producto más Vendidos
df_productosTop

producto  monto_venta
3         Auriculares     74175.58
5            Cafetera     59607.31
9   Cuadro decorativo     54297.60
16  Jarrón decorativo     51130.88
19    Lámpara de mesa     82276.38
20         Microondas     72562.89
25           Secadora     52115.45
27         Smartphone     54132.44

##***Actividad 3: Agregación***

 Resumir las ventas por categoría de producto y analizar los ingresos generados.

In [83]:
#creo un nuevo DataFrame agrupando los valores por categoría y sumando los montos de las ventas
df_ventas_Categoria = df_ventas.groupby(['categoria'], as_index=False)['monto_venta'].sum()

In [85]:
# corroboro la creación del nuevo DataFrame
df_ventas_Categoria

categoria  monto_venta
0         Decoración    479216.09
1  Electrodomésticos    505299.63
2        Electrónica    482577.80

In [94]:
# analizo el nuevo Dataframe y creo una nueva columna con el porcentaje de los montos de las ventas de cada categoría sobre el total
df_ventas_Categoria['porcentaje'] = df_ventas_Categoria['monto_venta']/df_ventas_Categoria['monto_venta'].sum()*100

In [95]:
# corroboro nueva columna
df_ventas_Categoria

categoria  monto_venta  porcentaje
0         Decoración    479216.09   32.664318
1  Electrodomésticos    505299.63   34.442224
2        Electrónica    482577.80   32.893459

##***Actividad 4: Integración de Datos***

Combinar los sets de datos de ventas y marketing para obtener una visión más amplia de las tendencias.

###***Repaso de los campos de los DataFrame 'Ventas' y 'Marketing'***

1.   Elemento de la lista
2.   Elemento de la lista



In [103]:
# repaso las columnas del DataFrame Ventas
df_ventas.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2998 entries, 0 to 2997
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype         
---  ------       --------------  -----         
 0   id_venta     2998 non-null   int64         
 1   producto     2998 non-null   object        
 2   precio       2998 non-null   float64       
 3   cantidad     2998 non-null   float64       
 4   fecha_venta  2998 non-null   datetime64[ns]
 5   categoria    2998 non-null   object        
 6   monto_venta  2998 non-null   float64       
dtypes: datetime64[ns](1), float64(3), int64(1), object(2)
memory usage: 164.1+ KB


In [104]:
# repaso las columnas del DataFrame Marketin
df_marketing.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 90 entries, 0 to 89
Data columns (total 6 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   id_campanha   90 non-null     int64         
 1   producto      90 non-null     object        
 2   canal         90 non-null     object        
 3   costo         90 non-null     float64       
 4   fecha_inicio  90 non-null     datetime64[ns]
 5   fecha_fin     90 non-null     datetime64[ns]
dtypes: datetime64[ns](2), float64(1), int64(1), object(2)
memory usage: 4.3+ KB


Ambos DataFrame comparten la categoría *'producto'* la cual puede ser utilizada para agrupar los DataFrame

###***Análisis del campo 'producto' en los DataFrame 'Ventas' y 'Marketing'***

####***DataFrame 'Marketing'***

In [108]:
# cuantos valores diferentes tiene 'productos' en "Marketing"
df_marketing['producto'].nunique()

30

In [105]:
# vemos cuantas veces se repite cada paroducto
df_marketing['producto'].sort_values().value_counts()

producto
Adorno de pared           3
Alfombra                  3
Aspiradora                3
Auriculares               3
Batidora                  3
Cafetera                  3
Candelabro                3
Consola de videojuegos    3
Cortinas                  3
Cuadro decorativo         3
Cámara digital            3
Elementos de cerámica     3
Espejo decorativo         3
Freidora eléctrica        3
Heladera                  3
Horno eléctrico           3
Jarrón decorativo         3
Laptop                    3
Lavadora                  3
Lámpara de mesa           3
Microondas                3
Parlantes Bluetooth       3
Plancha de vapor          3
Proyector                 3
Rincón de plantas         3
Secadora                  3
SmartWatch                3
Smartphone                3
Tablet                    3
Televisor                 3
Name: count, dtype: int64

In [112]:
# ordenamos los registros de "Marketing" en base a producto
df_marketing.sort_values(['producto']).head(30)

id_campanha                producto  canal  costo fecha_inicio  fecha_fin
0            74         Adorno de pared     TV   4.81   2024-03-20 2024-05-03
8            44         Adorno de pared  Email   5.08   2024-04-13 2024-05-10
70           14         Adorno de pared   RRSS   4.16   2024-10-22 2024-12-21
84           28                Alfombra   RRSS   5.82   2024-11-27 2025-01-08
4            58                Alfombra  Email   4.25   2024-03-31 2024-05-05
41           88                Alfombra     TV   7.39   2024-07-25 2024-08-31
85           70              Aspiradora     TV   3.06   2024-12-13 2024-12-29
29           40              Aspiradora  Email   6.24   2024-06-24 2024-08-12
52           10              Aspiradora   RRSS   7.16   2024-08-19 2024-09-19
24           47             Auriculares  Email   4.19   2024-06-02 2024-07-22
35           17             Auriculares   RRSS   6.06   2024-07-15 2024-08-15
79           77             Auriculares     TV   4.99   2024-11-11 2024-12-06
74           27                Batidora   RRSS   4.52   2024-11-01 2024-12-08
56           87                Batidora     TV   6.28   2024-08-27 2024-10-11
7            57                Batidora  Email   4.48   2024-04-10 2024-06-08
28           49                Cafetera  Email   5.20   2024-06-20 2024-07-22
53           19                Cafetera   RRSS   5.70   2024-08-19 2024-09-15
48           79                Cafetera     TV   7.26   2024-08-13 2024-10-11
68            9              Candelabro   RRSS   5.32   2024-10-08 2024-11-16
20           69              Candelabro     TV   3.25   2024-05-29 2024-06-30
38           39              Candelabro  Email   6.08   2024-07-17 2024-08-18
12           50  Consola de videojuegos  Email   4.97   2024-05-03 2024-05-28
36           20  Consola de videojuegos   RRSS   3.16   2024-07-15 2024-08-03
75           80  Consola de videojuegos     TV   4.64   2024-11-01 2024-11-20
11           43                Cortinas  Email   5.67   2024-05-02 2024-06-17
60           13                Cortinas   RRSS   5.22   2024-09-16 2024-10-15
51           73                Cortinas     TV   6.47   2024-08-17 2024-09-26
78           61       Cuadro decorativo     TV   5.30   2024-11-05 2024-12-23
10            1       Cuadro decorativo   RRSS   5.27   2024-04-27 2024-06-04
50           31       Cuadro decorativo  Email   5.28   2024-08-15 2024-09-12

**Observación:**

Cada producto tiene 3 repeticiones correspondientes a cada tipo de canal

####***DataFrame 'Ventas'***

In [113]:
# cuantos valores diferentes tiene 'productos' en "Ventas"
df_ventas['producto'].nunique()

30

In [106]:
df_ventas['producto'].sort_values().value_counts()

producto
Lámpara de mesa           176
Auriculares               143
Microondas                135
Cafetera                  117
Rincón de plantas         101
Smartphone                101
Cuadro decorativo         100
Cortinas                  100
Espejo decorativo         100
Cámara digital            100
Alfombra                  100
Adorno de pared           100
Aspiradora                100
Batidora                  100
Televisor                 100
SmartWatch                100
Secadora                  100
Freidora eléctrica        100
Lavadora                  100
Laptop                    100
Horno eléctrico           100
Jarrón decorativo         100
Plancha de vapor          100
Proyector                  99
Consola de videojuegos     99
Elementos de cerámica      99
Parlantes Bluetooth        98
Tablet                     58
Heladera                   48
Candelabro                 24
Name: count, dtype: int64

ver costos vs ganancias

#**Anexo**

**Bases de datos utilizadas:**
- [clientes.csv](https://github.com/FedeM-Git/TT-2C2025-Data-Analitycs-Notebooks/blob/main/Pre-Entrega_Maune/datasets/clientes.csv)
- [marketing.csv](https://github.com/FedeM-Git/TT-2C2025-Data-Analitycs-Notebooks/blob/main/Pre-Entrega_Maune/datasets/marketing.csv)
- [ventas.csv](https://github.com/FedeM-Git/TT-2C2025-Data-Analitycs-Notebooks/blob/main/Pre-Entrega_Maune/datasets/ventas.csv)